### PR TITLE
[TPU] Support sliding window and logit soft capping in the paged attention kernel for TPU.

### PR DESCRIPTION
**Pull request by vanbasten23**
After I added the sliding window and logit soft-capping support to the paged attention Pallas kernel, this PR is intended to added the sliding window and logit soft-capping at the vLLM level, so that we can support gemma model.

Test plans:
- pytest -s -v /workspace/vllm/tests/v1/tpu/test_pallas.py
- pytest -v -s tests/entrypoints/llm/test_accuracy.py::test_lm_eval_accuracy_v1_engine 2>&1 | tee out.txt

The Gemma model that we care about are [google/gemma-3-27b-it](https://huggingface.co/google/gemma-3-27b-it) and [google/gemma-3-1b-it](https://huggingface.co/google/gemma-3-1b-it).

cc @miladm

### CODE REVIEW

**Comment by brittrock** [comment]
This is great @vanbasten23 thanks for piping this in! Would we be able to add a Gemma 3 [text-only] test in please? 

Cc @robertgshaw2-redhat @yarongmu-google @yaochengji

File: tests/entrypoints/llm/test_accuracy.py [hunk 1]
```diff
@@ -13,18 +13,25 @@
...
+    "google/gemma-3-1b-it",
+]
 NUM_CONCURRENT = 500
 TASK = "gsm8k"
 FILTER = "exact_match,strict-match"
 RTOL = 0.03
-EXPECTED_VALUE = 0.58
+EXPECTED_VALUES = {
+    "Qwen/Qwen2-1.5B-Instruct": 0.58,
+    "google/gemma-3-1b-it": 0.25,
```
Comment (yaochengji) [inline tests/entrypoints/llm/test_accuracy.py:26]: Have you compared this score with GPU?

Comment (vanbasten23) [reply tests/entrypoints/llm/test_accuracy.py:26]: Thanks for the suggestion! I ran on GPU and got the same accuracy (0.25) for gemma-3-1b (test output on GPU: https://gist.github.com/vanbasten23/eda147d28db5f11187178771372ee39f)

File: tests/v1/tpu/test_pallas.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        attn_impl.forward(
+            layer=layer,
+            query=query,
+            key=key,
+            value=value,
+            kv_cache=kv_cache,
+            attn_metadata=attn_metadata,
+        )
+
+        mock_ragged_paged_attention.assert_called_once_with(
```
Comment (yaochengji) [inline tests/v1/tpu/test_pallas.py:80]: Should we also check the result?

File: tests/v1/tpu/test_pallas.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        head_size=head_size,
+        scale=scale,
+        num_kv_heads=num_kv_heads,
+        alibi_slopes=None,
+        sliding_window=sliding_window,
+        kv_cache_dtype="auto",
+        logits_soft_cap=logits_soft_cap,
+        attn_type=AttentionType.DECODER,
+    )
+    mock_vmem_limit_bytes = 1024
```
[the remark this answers is not in the document]
Comment (vanbasten23) [reply tests/v1/tpu/test_pallas.py:31]: This PR just pipes the sliding_window and logits_soft_cap from PallasAttentionBackendImpl.forward to the kernel. So this test verifies the 2 parameters are passed correctly. mock_vmem_limit_bytes here doesn't matter.

Since the torch.ops.xla.ragged_paged_attention is patched, we don't need to check the result.

File: tests/v1/tpu/test_pallas.py (hunk 2 above)
Comment (vanbasten23) [reply tests/v1/tpu/test_pallas.py:80]: replied in https://github.com/vllm-project/vllm/pull/15732#discussion_r2023822143

**Review by yaochengji — APPROVED** [review]
LGTM, thanks!

Note that the CI is not green.

**Comment by DarkLight1337** [comment]
You shuold merge from main to fix docker build

File: tests/v1/tpu/test_pallas.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (NickLucche) [reply tests/v1/tpu/test_pallas.py:31]: can you comment this is just checking args and correctness is tested in torch xla/jax experimental?

**Review by NickLucche — APPROVED** [review]
Great job!
I think some of the commits need signing, please take a look

File: tests/v1/tpu/test_pallas.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vanbasten23) [reply tests/v1/tpu/test_pallas.py:31]: done

[the remark this answers is not in the document]
Comment (NickLucche) [reply tests/v1/tpu/test_pallas.py:31]: thanks!

**Comment by bvrockwell** [comment]
@yaochengji we shouldn't merge this until @vanbasten23 validates 27B as well. Let's wait for these results as well.

**Review by bvrockwell — CHANGES_REQUESTED** [review]
waiting for multi chip validation - thank you!

**Comment by vanbasten23** [comment]
> @yaochengji we shouldn't merge this until @vanbasten23 validates 27B as well. Let's wait for these results .

I tested the 27b and it fails:
```
MODEL=google/gemma-3-27b-it
VLLM_USE_V1=1 vllm serve $MODEL --seed 42 --disable-log-requests --gpu-memory-utilization 0.95 --max-num-batched-tokens 1024 --max-num-seqs 512 --tensor-parallel-size 4 --max-model-len 2048 &
VLLM_USE_V1=1 python3 benchmarks/benchmark_serving.py --model $MODEL  --dataset-name random --random-input-len 1800 --random-output-len 128
```
the failure is in https://gist.github.com/vanbasten23/637bd3af3c3946c00e6fb1ce27892e46 which fails at tpu_woker.py, outside the kernel.

But I don't think it should block this PR. This PR intends to add sliding window/logit soft-capping support.  @bvrockwell The gemma-3 27B failure is a separate issue and can be fixed in a later PR.

**Comment by bvrockwell** [comment]
> > @yaochengji we shouldn't merge this until @vanbasten23 validates 27B as well. Let's wait for these results .
> 
> I tested the 27b and it fails:
> 
> ```
> MODEL=google/gemma-3-27b-it
> VLLM_USE_V1=1 vllm serve $MODEL --seed 42 --disable-log-requests --gpu-memory-utilization 0.95 --max-num-batched-tokens 1024 --max-num-seqs 512 --tensor-parallel-size 4 --max-model-len 2048 &
> VLLM_USE_V1=1 python3 benchmarks/benchmark_serving.py --model $MODEL  --dataset-name random --random-input-len 1800 --random-output-len 128
> ```
> 
> the failure is in https://gist.github.com/vanbasten23/637bd3af3c3946c00e6fb1ce27892e46 which fails at tpu_woker.py, outside the kernel.
> 
> But I don't think it should block this PR. This PR intends to add sliding window/logit soft-capping support. @bvrockwell The gemma-3 27B failure is a separate issue and can be fixed in a later PR.

ok understood @vanbasten23 , looking forward to the multichip gemma 3 PR! Thanks for adding sliding window and logits soft capping.

**Review by bvrockwell — APPROVED** [review]

**Review by robertgshaw2-redhat — APPROVED** [review]

**Comment by vanbasten23** [comment]
hi @robertgshaw2-redhat , I resolved the merge conflicts but the auto-merge was canceled. Could you enable the auto-merge again? Thanks!

**Comment by mgoin** [comment]
FYI the TPU V1 test was failing when this was merged
